### PR TITLE
feat: copy to clipboard

### LINKED ISSUE #232: Copy  to clipboard suggested command
I sometimes have to copy the suggested command to run in a separate terminal tab. Because it's long lived, interactivity is not working or I just want to quickly edit it before running. Support for quick copy of command to clipboard when preview is displayed would make this way easier.

### PR DESCRIPTION
**Pull request by brayo-pip**
Closes #232

> [!IMPORTANT]
> Adds clipboard copy functionality with platform-specific handling, refactors utility functions, and integrates into shell and tmux modules.
> 
>   - **Clipboard Functionality**:
>     - Adds `set_copytext()` and `copy()` in `clipboard.py` for clipboard management.
>     - Supports Linux (`xclip`, `wl-copy`) and macOS (`pbcopy`).
>   - **Utility Functions**:
>     - Moves `get_system_distro()` and `get_installed_programs()` to `platform_utils.py` and `util.py`.
>     - Refactors `print_preview()` in `util.py` to support clipboard copying.
>   - **Integration**:
>     - Updates `execute_shell()` in `shell.py` and `execute_tmux()` in `tmux.py` to use `print_preview()` with clipboard copy option.
> 
> <sup>This description was created by </sup>[Ellipsis](https://www.ellipsis.dev?ref=ErikBjare%2Fgptme&utm_source=github&utm_medium=referral)<sup> for 28e11a45dddd63e0cd94a40a33f55c3cf41a0a8c. It will automatically update as commits are pushed.</sup>

### CODE REVIEW

**Comment by 0xbrayo** [comment]
Closes #232

**Comment by 0xbrayo** [comment]
Experimented with a few approaches, `ask_execute` does not have access to the `cmd` variable. Passing it would involve rewriting all instances that `confirmFunc` is used. This was the simpler way I found to do it. Assumes that the `print_preview` is always called before `ask_execute` which is should always be the case. What do you think?

**Comment by 0xbrayo** [comment]
Noticed that other confirmation dialogs such as overwrite file have the "copy" option.

File: pyproject.toml [hunk 1]
```diff
@@ -25,6 +25,7 @@ python-dotenv = "^1.0.0"
 rich = "^13.5.2"
 tabulate = "*"
 pick = "^2.2.0"
+pyperclip = "^1.8.2"
```
Comment (ErikBjare) [inline pyproject.toml]: pyperclip seems abandoned: https://github.com/asweigart/pyperclip/issues/250 (not an endorsement of the alternative in the issue)

pyperclip also has `pytest` in their non-dev deps, which I found strange and undesireable.

File: gptme/util.py [hunk 2]
```diff
@@ -131,6 +133,7 @@ def epoch_to_age(epoch):
 def print_preview(code: str, lang: str):  # pragma: no cover
     print()
     print("[bold white]Preview[/bold white]")
+    set_text(code)
```
Comment (ErikBjare) [inline gptme/util.py]: Won't really work for patches, but I guess that's ok for now.

File: poetry.lock
Comment (ErikBjare) [inline poetry.lock:1]: Please use `poetry lock --no-update` when adding dependencies, so you don't accidentally update the locks of others at the same time.

**Review by ErikBjare — COMMENTED** [review]
Looks good, nice workaround with the preview stuff, but some concerns about pyperclip.

**Comment by 0xbrayo** [comment]
~~I will try `clipboard` instead~~ Also not maintained. Will call the native copy commands directly.

File: gptme/tools/tmux.py [hunk 3]
```diff
@@ -157,7 +157,7 @@ def execute_tmux(
     assert not args
     cmd = code.strip()
 
-    print_preview(f"Command: {cmd}", "bash")
+    print_preview(f"Command: {cmd}", "bash", copy=True)
```
Comment (ErikBjare) [inline gptme/tools/tmux.py:160]: Nice

File: gptme/platform_utils.py [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+    return "Linux"
+
+
+@functools.lru_cache
+def get_installed_programs(candidates: tuple[str, ...]) -> set[str]:
+    installed = set()
+    for candidate in candidates:
+        if shutil.which(candidate) is not None:
+            installed.add(candidate)
+    return installed
```
Comment (ErikBjare) [inline gptme/platform_utils.py]: I don't think we need a new file for this, just drop it into `utils.py`

File: gptme/util.py [hunk 5]
```diff
@@ -148,9 +168,16 @@ def ask_execute(question="Execute code?", default=True) -> bool:  # pragma: no c
     termios.tcflush(sys.stdin, termios.TCIFLUSH)  # flush stdin
 
     choicestr = f"[{'Y' if default else 'y'}/{'n' if default else 'N'}]"
+    copystr = r"\[c] to copy" if copiable else ""
     answer = console.input(
-        f"[bold bright_yellow on red] {question} {choicestr} [/] ",
+        f"[bold bright_yellow on red] {question} {choicestr}{copystr} [/] ",
     )
+    if copiable and "c" in answer.lower():
```
Comment (ErikBjare) [inline gptme/util.py]: What's the point of `set_copiable` here? Can't we just:

```suggestion
    if copy and answer.strip() == "c":
```

File: gptme/util.py [hunk 6]
```diff
@@ -148,9 +168,16 @@ def ask_execute(question="Execute code?", default=True) -> bool:  # pragma: no c
...
+        if copy():
+            print("Copied to clipboard.")
+        if len(answer) > 1:
+            answer = answer.lower().strip("c")
```
Comment (ErikBjare) [inline gptme/util.py]: I don't get this part

File: gptme/util.py (hunk 5 above)
Comment (ErikBjare) [reply gptme/util.py]: Nvm, thought it was a part of the other function.

Comment (ErikBjare) [inline gptme/util.py]: It should still be:

```suggestion
    if copiable and "c" == answer.strip().lower():
```

File: gptme/util.py (hunk 6 above)
Comment (0xbrayo) [reply gptme/util.py]: I want to let the user type "yc" to copy and still execute the code. Or just c to only copy the code.

File: gptme/util.py (hunk 5 above)
Comment (0xbrayo) [reply gptme/util.py]: See previous comment. Although I don't think it's that common of a use case

**Comment by 0xbrayo** [comment]
@ellipsis-dev review

**Comment by 0xbrayo** [comment]
@ErikBjare ping

File: gptme/util.py [hunk 7]
```diff
@@ -309,3 +341,23 @@ def path_with_tilde(path: Path) -> str:
...
+        with open("/etc/os-release") as f:
+            for line in f:
+                matches = re.search(regex, line)
+                if matches:
+                    return matches.string[matches.start(1) : matches.end(1)]
+    return "Linux"
+
+
+@functools.lru_cache
+def get_installed_programs(candidates: tuple[str, ...]) -> set[str]:
```
[the remark this answers is not in the document]
Comment (ErikBjare) [reply gptme/util.py:346]: This is still true

File: gptme/clipboard.py [hunk 8]
```diff
@@ -0,0 +1,45 @@
...
+def copy() -> bool:
+    """return True if successful"""
+    from .util import get_installed_programs
+
+    global text
+    if platform.system() == "Linux":
+        # check if xclip or wl-clipboard is installed
+        installed = get_installed_programs(("xclip", "wl-copy"))
+        if "wl-copy" in installed:
+            output = subprocess.run(["wl-copy"], input=text, text=True)
```
[the remark this answers is not in the document]
Comment (ErikBjare) [reply gptme/clipboard.py]: This is probably a good idea

**Review by ErikBjare — COMMENTED** [review]
See the remaining ellipsis comments (should fix mypy + lint warnings), looks good otherwise. Haven't actually tested it.

File: gptme/util.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (0xbrayo) [reply gptme/util.py:346]: Must have messed up the rebase

File: gptme/util.py [hunk 9]
```diff
@@ -309,3 +341,23 @@ def path_with_tilde(path: Path) -> str:
     if path_str.startswith(home):
         return path_str.replace(home, "~", 1)
     return path_str
+
+def get_system_distro() -> str:
```
Comment (ErikBjare) [inline gptme/util.py]: This doesn't seem to be used

**Comment by ErikBjare** [comment]
Nice, merged!

**Comment by NikitaBeloglazov** [comment]
Actually, I don't understand, there is the reason for building something unknown if you could just take clipman (an alternative to pyperclip) which was mentioned above. Or is there something missing?

**Comment by ErikBjare** [comment]
@NikitaBeloglazov There is no "building something unknown", the called executables should exist on their respective platforms. 

I don't want to add another dependency for something so simple.

**Pull request merged**